File: ExtremeStudio/DockPanel Forms/MainForm Docks/ProjExplorerDock.vb
Imports System.IO
Imports ExtremeParser
Imports ExtremeCore
Imports ExtremeStudio.My.Resources

Public Class ProjExplorerDock

    Public Sub RefreshList()
        filesList.SelectedPath = filesList.SelectedPath
    End Sub

    Public Function IsProtected(pPath As String) As Boolean
        If pPath.StartsWith(Path.Combine(MainForm.CurrentProject.ProjectPath, "configs")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "gamemodes")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "pawno")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "pawno\libpawnc.dll")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "pawno\pawn.ico")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "pawno\pawnc.dll")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "pawno\pawncc.exe")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "pawno\include")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "plugins")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "extremeStudio.config")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "server.cfg")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "samp-server.exe")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "samp-npc.exe")) _
            Or pPath = (Path.Combine(MainForm.CurrentProject.ProjectPath, "announce.exe")) _
        Then
            Return True
        End If 
        Return False
    End Function

    Private Sub ProjExplorerDock_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        'Wont be called multiple times: 
        filesList.MainDir = MainForm.CurrentProject.ProjectPath
        filesList.SelectedPath = MainForm.CurrentProject.ProjectPath
    End Sub
    Private Sub RefreshToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles RefreshToolStripMenuItem.Click
        RefreshList()
    End Sub

    Private Sub filesList_MouseDown(sender As Object, e As MouseEventArgs) Handles filesList.MouseDown
        If e.Button = MouseButtons.Right Then
            filesList.SelectedIndex = filesList.IndexFromPoint(e.Location)
            If filesList.SelectedItem <> Nothing AndAlso IsProtected(Path.Combine(filesList.SelectedPath, filesList.SelectedItem)) Then
                DeleteToolStripMenuItem.Enabled = False
                RenameToolStripMenuItem.Enabled = False
            Else
                DeleteToolStripMenuItem.Enabled = True
                RenameToolStripMenuItem.Enabled = True
            End If
            mouseRightClick.Show(MousePosition)
        End If
    End Sub

    Private Sub filesList_FileSelected(sender As Object, fse As ExtremeCore.FileSelectEventArgs) Handles filesList.FileSelected
        Dim ext = Path.GetExtension(fse.FileName)

        'Do different stuff depending on file type.
        If ext = ".inc" Or ext = ".pwn" Then
            MainForm.OpenFile(fse.FileName)
        Else
            MainForm.ShowStatus(translations.ProjExplorerDock_filesList_FileSelected_FileNotSupported, 5000, False)
            Process.Start(New ProcessStartInfo With { .WorkingDirectory = Path.GetDirectoryName(fse.FileName), .FileName = fse.FileName })
        End If
    End Sub

    Private Sub RenameToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles RenameToolStripMenuItem.Click
        If Directory.Exists(filesList.SelectedFile) Then
            ResetFolder:
            Dim input As New AdvacnedInputBox(translations.ProjExplorerDock_RenameToolStripMenuItem_Click_EnterNewNameTitle, translations.ProjExplorerDock_RenameToolStripMenuItem_Click_PleaseEnterNewDirName)
            If input.ResResult = AdvacnedInputBox.ReturnValue.InputResultOk Then
                If FilenameIsOk(input.ResText) Then
                    Try
                        My.Computer.FileSystem.RenameDirectory(filesList.SelectedFile, input.ResText)
                    Catch ex As Exception
                        MsgBox(ex.Message, MsgBoxStyle.OkOnly Or MsgBoxStyle.Critical)
                    End Try
                    RefreshList()
                Else
                    GoTo ResetFolder
                End If
            End If

        ElseIf File.Exists(filesList.SelectedFile) Then
            ResetFile:
            Dim input As New AdvacnedInputBox(translations.ProjExplorerDock_RenameToolStripMenuItem_Click_EnterNewNameTitle, translations.ProjExplorerDock_RenameToolStripMenuItem_Click_PleaseEnterNameForFile, defaultText:= Path.GetFileName(filesList.SelectedFile))
            If input.ResResult = AdvacnedInputBox.ReturnValue.InputResultOk Then
                If FilenameIsOk(input.ResText) Then
                    Try
                        My.Computer.FileSystem.RenameFile(filesList.SelectedFile, input.ResText)
                    Catch ex As Exception
                        MsgBox(ex.Message, MsgBoxStyle.OkOnly Or MsgBoxStyle.Critical)
                    End Try
                    RefreshList()
                Else
                    GoTo ResetFile
                End If
            End If

        End If
    End Sub

    Private Sub DeleteToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles DeleteToolStripMenuItem.Click
        If Directory.Exists(filesList.SelectedFile) Then
            Directory.Delete(filesList.SelectedFile)
            RefreshList()

        ElseIf File.Exists(filesList.SelectedFile) Then
            File.Delete(filesList.SelectedFile)
            RefreshList()

        End If
    End Sub

    Private Sub NewDirectoryToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles NewDirectoryToolStripMenuItem.Click
        If Directory.Exists(filesList.SelectedFile) Then
            'Go inside the dir.
            filesList.SelectedPath = Path.Combine(filesList.SelectedPath, filesList.SelectedItem.ToString)
        End If

        ResetChoose:
            Dim input As New AdvacnedInputBox(translations.ProjExplorerDock_RenameToolStripMenuItem_Click_EnterNewNameTitle, translations.ProjExplorerDock_NewDirectoryToolStripMenuItem_Click_PleaseEnterNameForDir)
            If input.ResResult = AdvacnedInputBox.ReturnValue.InputResultOk Then
                If FilenameIsOk(input.ResText) Then
                    Try
                        My.Computer.FileSystem.CreateDirectory(Path.Combine(filesList.SelectedPath, input.ResText))
                    Catch ex As Exception
                        MsgBox(ex.Message, MsgBoxStyle.OkOnly Or MsgBoxStyle.Critical)
                    End Try
                    RefreshList()
                Else
                    GoTo ResetChoose
                End If
            End If
    End Sub

    Private Sub NewFileToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles NewFileToolStripMenuItem.Click
        If Directory.Exists(filesList.SelectedFile) Then
            'Go inside the dir.
            filesList.SelectedPath = Path.Combine(filesList.SelectedPath, filesList.SelectedItem.ToString)
        End If

        ResetChoose:
            Dim input As New AdvacnedInputBox(translations.ProjExplorerDock_RenameToolStripMenuItem_Click_EnterNewNameTitle, translations.ProjExplorerDock_NewFileToolStripMenuItem_Click_PleaseEnterNameForFile, defaultText:= "example.pwn")
            If input.ResResult = AdvacnedInputBox.ReturnValue.InputResultOk Then
                If FilenameIsOk(input.ResText) Then
                    Try
                        My.Computer.FileSystem.WriteAllText(Path.Combine(filesList.SelectedPath, input.ResText), "", False)
                    Catch ex As Exception
                        MsgBox(ex.Message, MsgBoxStyle.OkOnly Or MsgBoxStyle.Critical)
                    End Try
                    RefreshList()
                Else
                    GoTo ResetChoose
                End If
            End If
    End Sub

    Private Sub filesList_KeyDown(sender As Object, e As KeyEventArgs) Handles filesList.KeyDown
        If e.KeyCode = Keys.Delete Then
            If IsProtected(Path.Combine(filesList.SelectedPath, filesList.SelectedItem)) = False Then
                DeleteToolStripMenuItem_Click(Me, Nothing)
            End If
        End If
    End Sub
End Class

File: ExtremeStudio/DockPanelForms/MainFormDocks/ProjExplorerDock.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Windows.Forms;
using ExtremeCore.Classes;
using ExtremeCore.Controls_And_Tools;
using Resources;
using WeifenLuo.WinFormsUI.Docking;

namespace ExtremeStudio.DockPanelForms.MainFormDocks
{
    public partial class ProjExplorerDock : DockContent
    {
        public ProjExplorerDock()
        {
            InitializeComponent();
        }

        public void RefreshList()
        {
            if(Visible)
                filesList.SelectedPath = filesList.SelectedPath;
        }

        private void ProjExplorerDock_Load(object sender, EventArgs e)
        {
            //Wont be called multiple times:
            filesList.MainDir = Program.MainForm.CurrentProject.ProjectPath;
            filesList.SelectedPath = Program.MainForm.CurrentProject.ProjectPath;
            filesList.IgnoredNames = new List<string>()
            {
                "configs",
                "dependencies",
                "extremeStudio.config",
                "pawn.json",
            };
            RefreshList();
        }

        private void RefreshToolStripMenuItem_Click(object sender, EventArgs e)
        {
            RefreshList();
        }

        private void filesList_MouseDown(object sender, MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Right)
            {
                filesList.SelectedIndex = filesList.IndexFromPoint(e.Location);
                if (filesList.SelectedItem != null)
                {
                    DeleteToolStripMenuItem.Enabled = false;
                    RenameToolStripMenuItem.Enabled = false;
                }
                else
                {
                    DeleteToolStripMenuItem.Enabled = true;
                    RenameToolStripMenuItem.Enabled = true;
                }

                mouseRightClick.Show(MousePosition);
            }
        }

        private void filesList_FileSelected(object sender, FileSelectEventArgs fse)
        {
            var ext = Path.GetExtension(Convert.ToString(fse.FileName));

            //Do different stuff depending on file type.
            if (ext == ".inc" || ext == ".pwn")
            {
                Program.MainForm.OpenFile(fse.FileName);
            }
            else
            {
                Program.MainForm.ShowStatus(translations.ProjExplorerDock_filesList_FileSelected_FileNotSupported, 5000, false);
                Process.Start(new ProcessStartInfo
                {
                    WorkingDirectory = Path.GetDirectoryName(Convert.ToString(fse.FileName)),
                    FileName = fse.FileName
                });
            }
        }

        private void RenameToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (Directory.Exists(Convert.ToString(filesList.SelectedFile)))
            {
                ResetFolder:
                AdvancedInputBox input = new AdvancedInputBox(
                    translations.ProjExplorerDock_RenameToolStripMenuItem_Click_EnterNewNameTitle,
                    translations.ProjExplorerDock_RenameToolStripMenuItem_Click_PleaseEnterNewDirName);
                if (input.ResResult == AdvancedInputBox.ReturnValue.InputResultOk)
                {
                    if (input.ResText.IsValidFileName())
                    {
                        try
                        {
                            Directory.Move(
                                Convert.ToString(filesList.SelectedFile), Path.Combine(Path.GetDirectoryName(filesList.SelectedFile), input.ResText)); 
                        }
                        catch (Exception ex)
                        {
                            MessageBox.Show(ex.Message, "", MessageBoxButtons.OK);
                        }

                        RefreshList();
                    }
                    else
                    {
                        goto ResetFolder;
                    }
                }

            }
            else if (File.Exists(Convert.ToString(filesList.SelectedFile)))
            {
                ResetFile:
                AdvancedInputBox input = new AdvancedInputBox(
                    translations.ProjExplorerDock_RenameToolStripMenuItem_Click_EnterNewNameTitle,
                    translations.ProjExplorerDock_RenameToolStripMenuItem_Click_PleaseEnterNameForFile,
                    Path.GetFileName(Convert.ToString(filesList.SelectedFile)));
                if (input.ResResult == AdvancedInputBox.ReturnValue.InputResultOk)
                {
                    if (input.ResText.IsValidFileName())
                    {
                        try
                        {
                            File.Move(
                                Convert.ToString(filesList.SelectedFile), Path.Combine(Path.GetDirectoryName(filesList.SelectedFile), input.ResText));

                            //If done, change path in openend documents
                            foreach (var doc in Program.MainForm.MainDock.Documents)
                            {
                                var oy = (EditorDock.EditorDock)doc.DockHandler.Form;
                                if ((string) oy.Editor.Tag == filesList.SelectedFile)
                                {
                                    oy.Editor.Tag = Path.Combine(Path.GetDirectoryName(filesList.SelectedFile),
                                        input.ResText);
                                    oy.TabText = input.ResText;
                                    oy.Text = input.ResText;
                                }
                            }
                        }
                        catch (Exception ex)
                        {
                            MessageBox.Show(ex.Message);
                        }

                        RefreshList();
                    }
                    else
                    {
                        goto ResetFile;
                    }
                }

            }
        }

        private void DeleteToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (MessageBox.Show(translations.ProjExplorerDock_DeleteToolStripMenuItem_Click_AreYouSure, "", MessageBoxButtons.YesNo,
                    MessageBoxIcon.Question) == DialogResult.No)
                return;

            if (Directory.Exists(Convert.ToString(filesList.SelectedFile)))
            {
                Directory.Delete(Convert.ToString(filesList.SelectedFile));
                RefreshList();

            }
            else if (File.Exists(Convert.ToString(filesList.SelectedFile)))
            {
                File.Delete(Convert.ToString(filesList.SelectedFile));
                RefreshList();

            }
        }

        private void NewDirectoryToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (Directory.Exists(Convert.ToString(filesList.SelectedFile)))
            {
                //Go inside the dir.
                filesList.SelectedPath = Path.Combine(Convert.ToString(filesList.SelectedPath),
                    Convert.ToString(filesList.SelectedItem.ToString()));
            }

            ResetChoose:
            AdvancedInputBox input = new AdvancedInputBox(
                translations.ProjExplorerDock_RenameToolStripMenuItem_Click_EnterNewNameTitle,
                translations.ProjExplorerDock_NewDirectoryToolStripMenuItem_Click_PleaseEnterNameForDir);
            if (input.ResResult == AdvancedInputBox.ReturnValue.InputResultOk)
            {
                if (GeneralFunctions.IsValidFileName(input.ResText))
                {
                    try
                    {
                        Directory.CreateDirectory(
                            Path.Combine(Convert.ToString(filesList.SelectedPath),
                                Convert.ToString(input.ResText)));
                    }
                    catch (Exception ex)
                    {
                        MessageBox.Show(ex.Message);
                    }

                    RefreshList();
                }
                else
                {
                    goto ResetChoose;
                }
            }
        }

        private void NewFileToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (Directory.Exists(Convert.ToString(filesList.SelectedFile)))
            {
                //Go inside the dir.
                filesList.SelectedPath = Path.Combine(Convert.ToString(filesList.SelectedPath),
                    Convert.ToString(filesList.SelectedItem.ToString()));
            }

            ResetChoose:
            AdvancedInputBox input = new AdvancedInputBox(
                translations.ProjExplorerDock_RenameToolStripMenuItem_Click_EnterNewNameTitle,
                translations.ProjExplorerDock_NewFileToolStripMenuItem_Click_PleaseEnterNameForFile, "example.pwn");
            if (input.ResResult == AdvancedInputBox.ReturnValue.InputResultOk)
            {
                if (GeneralFunctions.IsValidFileName(input.ResText))
                {
                    try
                    {
                        File.WriteAllText(
                            Path.Combine(Convert.ToString(filesList.SelectedPath),
                                Convert.ToString(input.ResText)), "");
                    }
                    catch (Exception ex)
                    {
                        MessageBox.Show(ex.Message);
                    }

                    RefreshList();
                }
                else
                {
                    goto ResetChoose;
                }
            }
        }

        private void filesList_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Delete)
            {
                DeleteToolStripMenuItem_Click(this, null);
            }
        }
    }
}